src/predictor.rs

```diff
@@ -0,0 +1,659 @@
+
+impl RevPredict for NoPredictor {
+    fn rev_predict_fix_endianness(
+        &self,
+        buffer: Bytes,
+        predictor_info: &PredictorInfo,
+        _: u32,
+        _: u32,
+    ) -> AsyncTiffResult<Bytes> {
+        let mut res = BytesMut::from(buffer);
+        fix_endianness(
+            &mut res[..],
+            predictor_info.endianness,
+            predictor_info.bits_per_sample[0],
+        );
+        Ok(res.into())
+    }
+}
+
+/// reverse horizontal predictor
+#[derive(Debug)]
+pub struct RevHorizontalPredictor;
+
+impl RevPredict for RevHorizontalPredictor {
+    fn rev_predict_fix_endianness(
+        &self,
+        buffer: Bytes,
+        predictor_info: &PredictorInfo,
+        tile_x: u32,
+        _: u32,
+    ) -> AsyncTiffResult<Bytes> {
+        let output_row_stride = predictor_info.output_row_stride(tile_x)?;
+        let samples = predictor_info.samples_per_pixel as usize;
+        let bit_depth = predictor_info.bits_per_sample[0];
+
+        let mut res = BytesMut::from(buffer);
+        fix_endianness(&mut res[..], predictor_info.endianness, bit_depth);
+        for buf in res.chunks_mut(output_row_stride) {
+            rev_hpredict_nsamp(buf, bit_depth, samples);
+        }
+        Ok(res.into())
+    }
+}
+
+/// Reverse predictor convenienve function for horizontal differencing
+///
+/// From image-tiff
+///
+/// This should be used _after_ endianness fixing
+pub fn rev_hpredict_nsamp(buf: &mut [u8], bit_depth: u16, samples: usize) {
+    match bit_depth {
+        0..=8 => {
+            for i in samples..buf.len() {
+                buf[i] = buf[i].wrapping_add(buf[i - samples]);
+            }
+        }
+        9..=16 => {
+            for i in (samples * 2..buf.len()).step_by(2) {
+                let v = u16::from_ne_bytes(buf[i..][..2].try_into().unwrap());
+                let p = u16::from_ne_bytes(buf[i - 2 * samples..][..2].try_into().unwrap());
+                buf[i..][..2].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        17..=32 => {
+            for i in (samples * 4..buf.len()).step_by(4) {
+                let v = u32::from_ne_bytes(buf[i..][..4].try_into().unwrap());
+                let p = u32::from_ne_bytes(buf[i - 4 * samples..][..4].try_into().unwrap());
+                buf[i..][..4].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        33..=64 => {
+            for i in (samples * 8..buf.len()).step_by(8) {
+                let v = u64::from_ne_bytes(buf[i..][..8].try_into().unwrap());
+                let p = u64::from_ne_bytes(buf[i - 8 * samples..][..8].try_into().unwrap());
+                buf[i..][..8].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        _ => {
+            unreachable!("Caller should have validated arguments. Please file a bug.")
+        }
+    }
+}
+
+/// Fix endianness. If `byte_order` matches the host, then conversion is a no-op.
```

Comment:
It's unclear from reading this that does not copy data.

If it is truly intended to be a no-op, seems better to do 

```rs
    #[cfg(target_endian = "little")]
	Endianness::LittleEndian => {},
	#[cfg(target_endian = "big")]
	Endianness::LittleEndian => match bit_depth ...
```
Then it's more clear that nothing is happening when the endianness of the data matches the host system's endianness

Reply:
this was directly copied from `image-tiff`. will do!